deployment/terraform/collect.go

```diff
@@ -0,0 +1,239 @@
+package terraform
+
+import (
+	"archive/tar"
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/mattermost/mattermost-load-test-ng/deployment/terraform/ssh"
+)
+
+type collectInfo struct {
+	instance string
+	src      string
+	compress bool
+}
+
+type file struct {
+	name string
+	data []byte
+}
+
+func saveCollection(files []file) error {
+	u, err := user.Current()
+	if err != nil {
+		return err
+	}
+
+	uid, err := strconv.ParseInt(u.Uid, 10, 32)
+	if err != nil {
+		return err
+	}
+
+	gid, err := strconv.ParseInt(u.Gid, 10, 32)
+	if err != nil {
+		return err
+	}
+
+	name := fmt.Sprintf("collection_%d", time.Now().Unix())
+	filename := fmt.Sprintf("./%s.tar", name)
+	f, err := os.Create(filename)
+	if err != nil {
+		return fmt.Errorf("failed to create file %w", err)
+	}
+	defer f.Close()
+
+	tw := tar.NewWriter(f)
+	hdr := &tar.Header{
+		Name:    name + "/",
+		Mode:    0755,
+		Uid:     int(uid),
+		Gid:     int(gid),
+		ModTime: time.Now(),
+	}
+	if err := tw.WriteHeader(hdr); err != nil {
+		return err
+	}
+
+	for _, file := range files {
+		hdr = &tar.Header{
+			Name:    name + "/" + file.name,
+			Mode:    0600,
+			Size:    int64(len(file.data)),
+			Uid:     int(uid),
+			Gid:     int(gid),
+			ModTime: time.Now(),
+		}
+
+		if err := tw.WriteHeader(hdr); err != nil {
+			return err
+		}
+
+		if _, err := tw.Write(file.data); err != nil {
+			return err
+		}
+	}
+
+	if err := tw.Close(); err != nil {
+		return err
+	}
+
+	fmt.Printf("%s generated.\n", filename)
+
+	return nil
+}
+
+func (t *Terraform) Collect() error {
+	output, err := t.Output()
+	if err != nil {
+		return err
+	}
+
+	if output.IsEmpty() {
+		return errors.New("no active deployment found")
+	}
+
+	if err := t.preFlightCheck(); err != nil {
+		return err
+	}
+
+	extAgent, err := ssh.NewAgent()
+	if err != nil {
+		return err
+	}
+
+	clients := make(map[string]*ssh.Client)
+	if output.HasProxy() {
+		sshc, err := extAgent.NewClient(output.Proxy.Value[0].PublicIP)
+		if err != nil {
+			return fmt.Errorf("error in getting ssh connection %w", err)
+		}
+		clients["proxy"] = sshc
+	}
+
+	for i, instance := range output.Instances.Value {
+		sshc, err := extAgent.NewClient(instance.PublicIP)
+		if err != nil {
+			return fmt.Errorf("error in getting ssh connection %w", err)
+		}
+		clients[fmt.Sprintf("app%d", i)] = sshc
+	}
+
+	for i, agent := range output.Agents.Value {
+		sshc, err := extAgent.NewClient(agent.PublicIP)
+		if err != nil {
+			return fmt.Errorf("error in getting ssh connection %w", err)
+		}
+		clients[fmt.Sprintf("agent%d", i)] = sshc
+		if i == 0 {
+			clients["coordinator"] = sshc
+		}
+	}
+
```

Comment:
Overall looks great. But I have a slight concern, which you would have probably seen it coming :)

I think we are adding too much functionality into the `terraform` package. The part after this line (L140) is not related to terraform at all, and can be in its own separate package (we can call it debug, 0/5 on it). Same for the `collection` function.

All that is needed from the terraform package is the `clients` map which can be done in the cmd/ltctl package itself IMO. It's just some wrapper code over the output and creating ssh clients. And then it can pass it to the `debug` package and let all of this be there. In future, this can also be reused by other deployment environments.

This is kinda what we did with the ssh package too. I think initially it was inside terraform and then moved to its own subpackage. I am okay if you want to do it later, but just wanted to open it up for discussion.

Thoughts ?

Reply:
That's all true. The reasoning was that the `ssh` package is actually used in many places while this is a standalone thing. Maybe for now I could move all non terraform functionality directly into a `cmd/ltctl/collect.go`.

Reply:
That sounds good to me. :+1: